src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -16,40 +16,69 @@
+   * equal to the other.
+   * Secondly, more the number of buckets, better the parallelism achieved during join in general
+   * assuming there is no resource constraint.
+   *
+   * If hybridScanEnabled is true, rank algorithm follows the algorithm above, but we prioritize
+   * the index with larger and usable index data for each index, so as to minimize the amount of
+   * data for on-the-fly shuffle or merge.
    *
-   * @param indexPairs index pairs for left and right side of the join. All index pairs are
+   * @param indexPairs Index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
-   * @return rearranged index pairs according to their ranking. The first is the best.
+   * @param hybridScanEnabled HybridScan config.
+   * @return Rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
-      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
-
+      leftPlan: LogicalPlan,
+      rightPlan: LogicalPlan,
+      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)],
+      hybridScanEnabled: Boolean): Seq[(IndexLogEntry, IndexLogEntry)] = {
     indexPairs.sortWith {
       case ((left1, left2), (right1, right2)) =>
+        // These common bytes was calculated and tagged in getCandidateIndexes.
+        // The value is the summation of common source files of the given plan and each index.
+        lazy val leftCommonBytes = left1
+          .getTagValue(leftPlan, IndexLogEntryTags.COMMON_BYTES)
+          .get + left2.getTagValue(leftPlan, IndexLogEntryTags.COMMON_BYTES).get
+        lazy val rightCommonBytes = right1
+          .getTagValue(rightPlan, IndexLogEntryTags.COMMON_BYTES)
+          .get + right2.getTagValue(rightPlan, IndexLogEntryTags.COMMON_BYTES).get
+
         if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
-          left1.numBuckets > right1.numBuckets
+          if (!hybridScanEnabled || (leftCommonBytes == rightCommonBytes)) {
+            left1.numBuckets > right1.numBuckets
+          } else {
+            // If both index pairs have the same number of buckets and Hybrid Scan is enabled,
+            // pick the pair with more common bytes with the given source plan, so as to
+            // reduce the overhead from handling appended and deleted files.
+            leftCommonBytes > rightCommonBytes
+          }
         } else if (left1.numBuckets == left2.numBuckets) {
+          // If one index pair has the same number of buckets, pick the one to avoid shuffling.
+          // This condition is prior to "more common bytes" condition for Hybrid Scan.
           true
         } else if (right1.numBuckets == right2.numBuckets) {
           false
         } else {
-          true
+          // Pick the pair with "more common bytes" if both pairs have different number of buckets.
+          !hybridScanEnabled || leftCommonBytes > rightCommonBytes
```

Comment:
Also in this scenario, one side is being shuffled right?
Consider this scenario:
we choose left1, left2 because left1.commonBytes + left2.commonBytes > right1.commonBytes + right2.commonBytes

It is possible that left1.commonBytes = 10, left2.commonBytes = 1000. right1.commonBytes = 500, right2.commonBytes = 500

But we choose left1 and left2. Now it is possible that left2 gets reshuffled because #buckets is different. But if we chose right1 and right2, we would get less shuffling right?

Reply:
Oh, I thought both side will be shuffled with Spark's config value.
In that case, right1 + right2 should be prior. Then we could do:
`min(left1.commonBytes, left2.commonBytes) > min(rigtht1.commonBytes, right2.commonBytes)`, right?

Reply:
actually it's slightly more different.

consider `t1 Join t2 where ...`
 - if t1.buckets == t2.buckets => no shuffling
 - if t1.buckets != t2.buckets != spark.sql.shuffle.partitions => both sides get shuffled
 - if t1.buckets != t2.buckets but t1.buckets == spark.sql.shuffle.partitions => only t2 gets shuffled
 - if t1.buckets != t2.buckets but t2.buckets == spark.sql.shuffle.partitions => only t1 gets shuffled